lib/elixir/lib/system.ex

```diff
@@ -761,22 +767,7 @@ defmodule System do
   """
   @spec put_env(Enumerable.t()) :: :ok
   def put_env(enum) do
-    Enum.each(enum, fn
-      {key, nil} ->
-        :os.unsetenv(to_charlist(key))
-
-      {key, val} ->
-        key = to_charlist(key)
-
-        case :string.find(key, "=") do
-          :nomatch ->
-            :os.putenv(key, to_charlist(val))
-
-          _ ->
-            raise ArgumentError,
-                  "cannot execute System.put_env/1 for key with \"=\", got: #{inspect(key)}"
-        end
-    end)
+    Enum.each(enum, fn {key, value} -> put_env(to_string(key), value) end)
```

Comment:
put_env/1 can still more than put_env/2. It can take arbitrary keys and values. Now we convert `key` to string and then to charlist, which is not optimal. I could make put_env/1 accept more than strings but I don't feel like this is going in a good direction

Reply:
Yeah, I would leave the previous implementation as is then. The error message will be always precise. The only change to this function is to also use `:binary.match`.

Reply:
I belive there is `:string.find` as we work on a charlist. In other case we would have to convert to string, do :binary.match and then convert to charlist which is the same as in the current version